addons/mrp_account/models/product.py

```diff
@@ -47,20 +48,11 @@ def _compute_average_price(self, qty_invoiced, qty_to_invoice, stock_moves):
         bom = self.env['mrp.bom']._bom_find(product=self, company_id=stock_moves.company_id.id, bom_type='phantom')
         if not bom:
             return super()._compute_average_price(qty_invoiced, qty_to_invoice, stock_moves)
-        dummy, bom_lines = bom.explode(self, 1)
-        bom_lines = {line: data for line, data in bom_lines}
         value = 0
-        for move in stock_moves:
-            if move.state == 'cancel':
-                continue
-            bom_line = move.bom_line_id
-            if bom_line in bom_lines:
-                bom_line_data = bom_lines[bom_line]
-                line_qty = bom_line.product_uom_id._compute_quantity(bom_line_data['qty'], bom_line.product_id.uom_id)
-            else:
-                # bom was altered (i.e. bom line removed) after being used
-                line_qty = move.product_qty
-            value += line_qty * move.product_id._compute_average_price(qty_invoiced * line_qty, qty_to_invoice * line_qty, move)
+        for bom_line, moves_list in groupby(stock_moves.filtered(lambda sm: sm.state != 'cancel'), lambda sm: sm.bom_line_id):
+            line_qty = bom_line.product_uom_id._compute_quantity(bom_line.product_qty, bom_line.product_id.uom_id)
```

Comment:
```suggestion
            if not bom_line:
               for move in moves_list:
                   value += move.product_qty * move.product_id._compute_average_price(qty_invoiced * move.product_qty, qty_to_invoice * move.product_qty, move)
                   continue
            line_qty = bom_line.product_uom_id._compute_quantity(bom_line.product_qty, bom_line.product_id.uom_id)
```
To handle previous case